torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+                assert data_ptr == new_inputs[idx].data_ptr()
+            else:
+                dst = self._reconstruct_from_tensor_metadata(
+                    self.non_static_inputs_metadata[idx], storage_cache
+                )
+                src = new_inputs[idx]
+                expanded_dims = self.expanded_dims[idx]
+
+                dst = index_expanded_dims(dst, expanded_dims)
+                src = index_expanded_dims(src, expanded_dims)
+                # TODO - one jit kernel across multiple inputs
+                dst.copy_(src)
+
+        new_inputs.clear()
+        self.graph.replay()
+
+        # outputs is not None on first execution
+        if self.recording_outputs is not None:
+            outputs = self.recording_outputs
+            self.recording_outputs = None
+            self._add_first_outputs(outputs)
+
+            return outputs
+
+        outputs = [
+            self._reconstruct_from_tensor_metadata(metadata, storage_cache)
+            for metadata in self.outputs_metadata
+        ]
+
+        self._add_replayed_outputs(outputs)
+
+        return outputs
+
+    def all_outputs_are_dead(self):
+        "All outputs of the path from this node to its root are dead"
+        for i, j in self.live_indices_after_graph:
+            if is_live(self.path_weakrefs[i][j]):
+                return False
+
+        return True
+
+    def _warmup(self, model, stream, inps):
+        "Warmup the model"
+        # TODO - optimize memory of warmup (deallocate previous inputs, re-use existing memory for running kernels)
+        torch.cuda.synchronize()
+        stream.wait_stream(torch.cuda.current_stream())
+        # copy inputs because list will get cleared in model invocation
+        with torch.cuda.stream(stream):
+            model(list(inps))
+        stream.synchronize()
+        torch.cuda.current_stream().wait_stream(stream)
+        torch.cuda.synchronize()
+
+    def _record(self, model, stream, inputs):
+        "Record the model"
+        with torch.cuda.graph(self.graph, stream=stream, pool=self.cuda_graphs_pool):
+            static_outputs = model(inputs)
+
+        # running model should reclaim memory
+        assert len(inputs) == 0
+
+        if not isinstance(static_outputs, (list, tuple)):
+            static_outputs = (static_outputs,)
+
+        return static_outputs
+
+    def _add_first_outputs(self, outputs):
+        "Add the outputs from the first invocation of the node and set up metadata"
+        prev_liveness = self.recorded_liveness_before_graph
+        curr_liveness = self._get_liveness(self.path_weakrefs)
+
+        delta = self._get_different_indices(prev_liveness, curr_liveness)
+        self.expected_dead_indices_after_graph = delta
+
+        assert len(self.outputs_weakrefs) == 0
+        weak_refs = [self._map_to_ref(o) for o in outputs]
+        self.outputs_weakrefs.extend(weak_refs)
+        self.recorded_liveness_after_graph = self._get_liveness(self.path_weakrefs)
```

Comment:
did `self.path_weakrefs` change between here and line 419?

Reply:
Yes, `self.outputs_weakrefs.extend(weak_refs)` would have changed it. i'll add comment